core/trino-main/src/main/java/io/trino/sql/planner/rowpattern/IrRowPatternToProgramRewriter.java

```diff
@@ -0,0 +1,278 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.sql.planner.rowpattern;
+
+import com.google.common.collect.ImmutableList;
+import io.trino.operator.window.matcher.Done;
+import io.trino.operator.window.matcher.Instruction;
+import io.trino.operator.window.matcher.Jump;
+import io.trino.operator.window.matcher.MatchEnd;
+import io.trino.operator.window.matcher.MatchLabel;
+import io.trino.operator.window.matcher.MatchStart;
+import io.trino.operator.window.matcher.Program;
+import io.trino.operator.window.matcher.Save;
+import io.trino.operator.window.matcher.Split;
+
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.IntStream;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.Collections2.orderedPermutations;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class IrRowPatternToProgramRewriter
```

Comment:
This class should be in the execution package as it relates to the translation between the IR and what's needed by the operator